Ub vs tbd (#10)
nikomatsakis/project-ffi-unwind: spec-terminology.md

# Terminology about specifications

Language and platform specifications have several different terms used
to describe how well-defined a language feature is, i.e., how well
constrained the runtime behavior is. In cases where our terminology
overlaps with that from other communities, we try to remain generally
compatible.

<a name="UB"></a>

## Undefined Behavior

> As is typical within the Rust community we use the phrase **undefined

Reviewer: I think it would be better to just link here to the definition in the UCG repo glossary.

behavior** to refer to illegal program actions that can result in
arbitrary results. In short, "undefined behavior" is always a bug and
never something you should do. See the [Rust
reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html)
for more details.

Our usage of the term is generally the same as the [standard
usage](https://en.wikipedia.org/wiki/Undefined_behavior) from other
languages.

<a name="LLVM-UB"></a>

### LLVM-undefined behavior (LLVM-UB)

> As a special case of undefined behavior, we use the phrase **LLVM

Reviewer: It is unclear whether LLVM-UB is also UB in Rust.

Author: I think that the phrase "special case" implies that it is. However, I'm not sure how important the distinction really is at the end of the day.

That said, I debated about whether to include that phrase. I think it's probably worth distinguishing LLVM-UB as an orthogonal concept -- LLVM-UB is whatever LLVM defines as UB. Any case where Rust generates LLVM that exhibits LLVM-UB is a bug unless that behavior is also UB in Rust -- but those bugs can and do happen so it's good to be able to talk about it.

Reviewer: > I think that the phrase "special case" implies that it is. 

I thought that the goal of the term "LLVM-UB" was to use it for those cases in which Rust has UB, but we guarantee that the generated LLVM-IR does not. For example, unwinding through "C unwind" is Rust-UB because we don't want to guarantee what that does, but it is not LLVM-UB because we want people to use it. If "LLVM-UB" implies "Rust-UB" then by saying that "C unwind" is not "LLVM-UB" we are saying that it is not "Rust-UB". Does that make sense ?

Author: I think this concern is basically resolved now -- I've clarified that LLVM-UB just refers to what is UB in LLVM, which is (hopefully) a subset of Rust UB (but not necessarily).

Moreover, as per this document, unwinding through "C unwind" is not Rust UB, but it *is* unspecified (or, more precisely, "TBD").

Author: Resolving.

compilation, a struct must have *some* layout. Moreover, we guarantee
that programs can (for example) use `sizeof` to determine the size of
that layout, or access fields using Rust syntax like `foo.bar`. This
requires the layout to be communicated in some fashion but doesn't
specify how that is done.

[ucg-struct]: https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/layout/structs-and-tuples.md

Our usage of the term is generally the same as the [standard
usage](https://en.wikipedia.org/wiki/Unspecified_behavior) from other
languages.

<a name="TBD"></a>

## To Be Defined Behavior (TBD)

We refer to some behavior as **to be defined** to indicate that --
while it is currently unspecified -- we *intend* to define that
behavior at some point as part of this project group (though plans can
change). This helps to define the scope of the group, but it also
indicates behavior that you would be able to rely upon in the future
in stable code. Note that TBD behavior is **still unspecified** until
a formal decision is made, though, so if you rely on it today, your
code cannot be considered stable Rust (even if it compiles on the

> stable compiler).

Reviewer: So my idea of TBD is that it is a term that is used within this WG only, where we use it to say that we haven't given the issue the necessary attention to specify things in any way but have to do so at some point. So in practice, it is just undefined behavior. 

I also think that TBD should not reach "Rust", that is, everything that's TBD needs to be specified to be some type of behavior before stabilization (including just saying that its UB). 

Why use it instead of just UB? As a placeholder to differentiate the UB that's "UB by design" from the UB that's "UB because we haven't given this any thought yet, but want to do so".

Author: > So my idea of TBD ...

I think this basically aligns with what I wrote. I can't tell if there is a suggested edit in there? If so, can you try to make it more concrete?

> I also think that TBD should not reach "Rust", that is, everything that's TBD needs to be specified to be some type of behavior before stabilization (including just saying that its UB).

You and I have a different opinion here. I agree that in general it would be *nice* if TBD behavior remains on nightly, but I think it is acceptable for it to hit stable as well, and this has already happened in all sorts of areas -- the most notable case is the interaction of raw pointers, but there are plenty of other examples (e.g., the details of symbol visibility and shraed libraries is something we've tinkered with a few times). In general, I think it is a strict improvement on the status quo to offer people a plan they can adopt and a clear definition of what they can and cannot rely upon (and make progress on the latter until it suffices). However, I would be happy to try and write-up a "decision report" and take it to the lang team for a more formal decision.

Reviewer: > You and I have a different opinion here. I agree that in general it would be nice if TBD behavior remains on nightly, but I think it is acceptable for it to hit stable as well, and this has already happened in all sorts of areas -- the most notable case is the interaction of raw pointers, but there are plenty of other examples (e.g., the details of symbol visibility and shraed libraries is something we've tinkered with a few times).

I think the issue is that by TBD you refer to what I refer to as "unspecified behavior", where we say that this behavior is defined to do something that's not UB, but we don't say what that something is.

Right now, almost everything is unspecified in these documents, but some things are "unspecified by design" - we have looked at them, and this is something that we have decided that we don't want to specify, like the Rust unwinding ABI - and some are "unspecified because we haven't had the time to look at it yet". 

I was hoping that TBD could be used for these second type of unspecified things, so that we would at least look at them before stabilization, and say whether we want them to be undefined, unspecified, or something else. 

Maybe from this point-of-view it is clearer why I think that shipping a TBD is not a great idea. It would mean that we have recognized that there is something important enough to mention here as something to consider, but that we haven't considered it enough to even say that we provide no guarantees about it.

Author: So I think the full range of things are:

* Undefined behavior -- always a bug. As we discussed some in privmsg, the "miri virtual machine" panics here.
* Unspecified behavior -- the machine does not panic, but we don't say what happens. As you point out, there are a number of *reasons* we might not have said what happens yet:
    * We never want to specify this because we expect to want to change it. Example: Rust struct layout (without a repr attribute).
    * We expect eventually to specify this, but we haven't figure out just how it should be specified *yet*. This is what I am calling TBD, and I think you are too.

An orthogonal discussion is whether "TBD" behavior should ever be possible on stable. It seems clear that this is *suboptimal*, in that it leaves room for "de facto" dependencies. 

I have tried to summarize this in a [hackmd doc](https://hackmd.io/7ZcKCaRiRO2_I_yK8yeauQ).

Author: Resolving this concern based on [this comment](https://rust-lang.zulipchat.com/#narrow/stream/210922-wg-ffi-unwind/topic/tbd.20vs.20ub.20vs.20unspecified.20behavior/near/178398565).

never something you should do. See the [Rust
reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html)
and the [entry in the Unsafe Code Guidelines glossary][ucg-ub] for more details.

[ucg-ub]: https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#undefined-behavior

Our usage of the term is generally the same as the [standard
usage](https://en.wikipedia.org/wiki/Undefined_behavior) from other
languages.

<a name="LLVM-UB"></a>

## LLVM-undefined behavior (LLVM-UB)

We use the phrase **LLVM undefined behavior** to indicate things that
are considered undefined behavior by LLVM itself. Barring bugs, the
Rust compiler should never produce LLVM IR that contains LLVM-UB
unless the behavior in question is *also* UB in Rust. Of course, there
are bugs in the Rust compiler from time to time, and hence it can
happen that we generate LLVM IR which contains LLVM-UB even if the
corresponding Rust source code is meant to be fully defined (see
e.g. [rust-lang/rust#28728]).  The main reason it is worth separating
LLVM-UB from the more general form of Rust UB is that, while both
forms of UB can cause arbitrary things to happen in your

> code. However, as a practical measure, LLVM-UB is much more *likely

Reviewer 2: Looks like these sentences are really supposed to be one sentence that was inadvertently split.

```suggestion
code, LLVM-UB is much more *likely* to do so in practice.
```

(Can't submit a multi-line suggestion yet, but of course this would mean deleting the next line 😄 )

to* in practice.

[rust-lang/rust#28728]: https://github.com/rust-lang/rust/issues/28728

<a name="unspecified"></a>

## Unspecified behavior

We use the term "unspecified behavior" to refer to behavior that may
vary across Rust releases, depending on what options are given to the
compiler, or even -- in extreme cases -- across executions of the Rust

> compiler. However, unlike undefined behavior, the resulting execution

Reviewer 2: It may be worth noting variation across platforms, too.
